encoder/src/main/java/openlr/encoder/routesearch/SecondShortestRouteChecker.java

```diff
@@ -0,0 +1,192 @@
+    }
+
+
+    /**
+     *
+     * @param location route
+     * @param relativeThreshold threshold relative to the route length
+     * @return Instance of SecondShortestRouteChecker for the location with threshold
+     * @throws OpenLREncoderProcessingException
+     */
+    public static SecondShortestRouteChecker on(List<? extends Line> location, double relativeThreshold) throws OpenLREncoderProcessingException {
+        if (location.isEmpty()) {
+            throw new OpenLREncoderProcessingException(OpenLREncoderProcessingException.EncoderProcessingError.NO_ROUTE_FOUND_ERROR);
+        }
+        GeoCoordinates destinationStart = location.get(location.size() - 1).getStartNode().getGeoCoordinates();
+        int locationLength = location.stream().mapToInt(Line::getLineLength).sum();
+
+        int maxLengthAllowed = locationLength + (int) (locationLength * relativeThreshold);
+
+        return new SecondShortestRouteChecker(destinationStart, location, maxLengthAllowed, relativeThreshold);
+    }
+
+    /**
+     *
+     * @param lengthAlongSecondShortestRoute length of the second shortest route
+     * @param lengthAlongLocation length of the route along location to which the second shortest route gives alternative to
+     * @return true: if the length of the second shortest route is not greater than route along location by relative threshold percentage
+     *         false: if the length of the second shortest route is greater than route along location by relative threshold percentage
+     */
+    boolean VerifyThreshold(int lengthAlongSecondShortestRoute, int lengthAlongLocation){
+        return (lengthAlongSecondShortestRoute < (lengthAlongLocation + (int) (lengthAlongLocation * relativeThreshold)));
+    }
+
+
+    /**
+     * @param length heuristic length
+     * @return true  if the heuristic is less than the maximum  length
+     *         false if the heuristic is greater than the maximum  length
+     */
+    private boolean networkLengthFilter(Integer length) {
+        return (length < maxLengthAllowed);
+    }
+
+    /**
+     * @param lastLineIndex index of the road segment in the route
+     * @return length along the location from start till the road segment at the lastLineIndex
+     */
+    private int lengthAlongLocation(int lastLineIndex) {
+        int routeLength = 0;
+        for (int index = 0; index <= lastLineIndex; ++index) {
+            routeLength += location.get(index).getLineLength();
+        }
+        return routeLength;
+    }
+
+    /**
+     *
+     * @param index index of the road segment in the location between the two location reference points
+     * @return true if an alternate path of length under the the relative threshold exist
+     *         false if no alternate path of length under the the relative threshold exist
+     */
+    public boolean hasValidDeviationBefore(int index) {
+        if (index > 0 && index < location.size() - 1) {
```

Comment:
you can replace this if..else with
`  if (index <= 0 ){ return false;}`
then the calculation logic.

Reply:
ac5407dc43f6f89b768b92e9571ed9f04942a5d2